Firestore/core/src/firebase/firestore/core/firestore_client.mm

```diff
@@ -0,0 +1,500 @@
+#import "Firestore/Source/API/FIRDocumentSnapshot+Internal.h"
+#import "Firestore/Source/API/FIRFirestore+Internal.h"
+#import "Firestore/Source/API/FIRQuery+Internal.h"
+#import "Firestore/Source/API/FIRQuerySnapshot+Internal.h"
+#import "Firestore/Source/API/FIRSnapshotMetadata+Internal.h"
+#import "Firestore/Source/Core/FSTSyncEngine.h"
+#import "Firestore/Source/Core/FSTView.h"
+#import "Firestore/Source/Local/FSTLRUGarbageCollector.h"
+#import "Firestore/Source/Local/FSTLevelDB.h"
+#import "Firestore/Source/Local/FSTLocalSerializer.h"
+#import "Firestore/Source/Local/FSTLocalStore.h"
+#import "Firestore/Source/Local/FSTMemoryPersistence.h"
+#import "Firestore/Source/Local/FSTPersistence.h"
+#import "Firestore/Source/Remote/FSTSerializerBeta.h"
+#import "Firestore/Source/Util/FSTClasses.h"
+
+#include "Firestore/core/src/firebase/firestore/api/settings.h"
+#include "Firestore/core/src/firebase/firestore/auth/credentials_provider.h"
+#include "Firestore/core/src/firebase/firestore/core/database_info.h"
+#include "Firestore/core/src/firebase/firestore/core/event_manager.h"
+#include "Firestore/core/src/firebase/firestore/model/database_id.h"
+#include "Firestore/core/src/firebase/firestore/model/document_set.h"
+#include "Firestore/core/src/firebase/firestore/model/mutation.h"
+#include "Firestore/core/src/firebase/firestore/remote/datastore.h"
+#include "Firestore/core/src/firebase/firestore/remote/remote_store.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/delayed_constructor.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+#include "Firestore/core/src/firebase/firestore/util/statusor.h"
+#include "Firestore/core/src/firebase/firestore/util/string_apple.h"
+#include "absl/memory/memory.h"
+
+namespace firebase {
+namespace firestore {
+namespace core {
+
+using firestore::Error;
+using api::DocumentReference;
+using api::DocumentSnapshot;
+using api::QuerySnapshot;
+using api::Settings;
+using api::SnapshotMetadata;
+using api::ThrowIllegalState;
+using auth::CredentialsProvider;
+using auth::User;
+using core::DatabaseInfo;
+using core::ListenOptions;
+using core::EventManager;
+using core::Query;
+using core::QueryListener;
+using core::ViewSnapshot;
+using local::LruParams;
+using model::DatabaseId;
+using model::Document;
+using model::DocumentKeySet;
+using model::DocumentMap;
+using model::MaybeDocument;
+using model::Mutation;
+using model::OnlineState;
+using remote::Datastore;
+using remote::RemoteStore;
+using util::Path;
+using util::AsyncQueue;
+using util::DelayedConstructor;
+using util::DelayedOperation;
+using util::Executor;
+using util::Status;
+using util::StatusCallback;
+using util::StatusOr;
+using util::StatusOrCallback;
+using util::TimerId;
+
```

Comment:
Regarding two-step initialization: how about hiding it behind a static factory function?
* the constructor will become private;
* instead, there will be a `Create` function that looks like:
```cpp
/*static*/ std::shared_ptr<FirestoreClient> FirestoreClient::Create(/* parameters here */) {
  auto result = std::make_shared<FirestoreClient>(/*arguments...*/);
  result->Initialize(settings);
  return result;
}
```
This would:
* hide the complexity;
* prevent errors in code that uses `FirestoreClient`;
* allow you to drop `VerifyInitialized` checks and `client_initialized_` member variable.

Reply:
You could go even a step further and move the code in `Initialize` into the factory function. That would remove the need for two step initialization altogether and then the platform-specific weirdness would just be that it's a factory function instead of a constructor.

Reply:
Done.